designs/opencl_kernel_generator.md

```diff
@@ -0,0 +1,84 @@
+- Feature Name: opencl_kernel_generator
+- Start Date: 2019-10-21
+- RFC PR: (leave this empty)
+- Stan Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Define operators and functions on matrix_cl that will use 
+[expresson templates](https://en.wikipedia.org/wiki/Expression_templates) 
+to construct expressions. When assigned to a matrix_cl an expression will generate an 
+OpenCL kernel an execute it to calculate the result of the expression.
+
+# Motivation
+[motivation]: #motivation
+
+Using expression templates is much simpler than writing OpenCL kernels by hand and it does 
+not require knowledge of OpenCL and parallel programming.
+
+Using one kernel for multiple operations is faster than using one kernel per operation. Each 
+kernel must load its arguments from global global GPU memory and store results back. Memory 
+transfers are relatively slow compared to calculations. In case of simple kernels majority 
+of kernel execution time is spent for memory transfers.  
+
+While not any kernel can be rewritten with kernel generator, many, if not most of ones 
+needed in Stan Math should qualify.  
+
+# Explanation
+[guide-level-explanation]: #explanation
+
+A kernel generator expression consists of operations, for example addition. While not 
+operations in mathematical sense, accesses to matrices and scalars are also operations 
+in kernel generator.
+
+Each operation is implemented as a templated class. When instantiated those templates are given
+types of the subexpressions that are arguments to the particular operation. User-facing 
+functions are used to simplify construction of operations, since class templates can not 
+be deduced in C++14. 
+
+Each operation object is responsible for generating kernel code for the operation it 
+represents. Each operation selects a unique variable name and generates kernel code that 
+declares a variable and assigns the result of the operation to it. If the operation has any
+arguments that are operations themselves it can access variables they use for their results.
+If the operation needs any other data it can specify kernel arguments. For example an operation 
+that accesses a matrix might need a pointer to matrix data and size of the matrix.
+
+Some operations can also be used in expressions that are assigned to. The most basic case 
+is access to matrix - i.e. storing the result of a kernel. Such operations can also generate
+kernel code for appearing on the left-hand-side of assignment. While the code is different the
+process of generating it is similar to generating code for an operation on the right-hand side 
+of an assignment.
+
+Kernel generator expressions can consist of operations, matrices and scalars. Operations have
+appropriate methods for constructing kernel source. Matrices and scalars on the other hand do 
+not. So they need to be wrapped in appropriate wrapper that is an operation whenever they are
+used in a kernel generator expression.
```

Comment:
Sorry, but it was not clear to me from the template which parts of the explanation should go into guide and which into reference. Since the example (logger) spec does not split explanation into two parts I thought it would be OK if I did the same.